### PR TITLE
Finite Volume PorousFlow

### PR DESCRIPTION
**Pull request by cpgr**
This is the initial work to support FV in the PorousFlow module.

@loganharbour, would you mind taking a look at this when you get chance before I go too far down the rabbit hole. 😆 I think this was the minimum I could do to get a working FV test problem while having all of the current tests in PorousFlow still working.

### CODE REVIEW

**Comment by cpgr** [comment]
Hmm, gcc doesn't like this. It looks like I need lots of `this->template`'s in the code.

**Comment by cpgr** [comment]
> Hmm, gcc doesn't like this. It looks like I need lots of `this->template`'s in the code.

Luckily not too many `this->templates` were required!

**Comment by mangerij** [comment]
Hi, I noticed Ferret failed here but when I looked at CIVET, I saw that all tests passed. Anyone know why?

**Comment by loganharbour** [comment]
> Hi, I noticed Ferret failed here but when I looked at CIVET, I saw that all tests passed. Anyone know why?

Unrelated issue. Should be good now.

> @loganharbour, would you mind taking a look at this when you get chance before I go too far down the rabbit hole. 

👍🏼 on my list.

File: large_media
Comment (cpgr) [reply large_media:1]: Not sure what happened here...

**Comment by cpgr** [comment]
I've just realised that it doesn't seem like there is any way to find the documentation (https://mooseframework.inl.gov/docs/PRs/21172/site/automatic_differentiation/templated_objects.html) I wrote. It follows on fromhttps://mooseframework.inl.gov/docs/PRs/21172/site/automatic_differentiation/ but I can't find out where this links from. It should possibly be listed in 'Application development'?

**Comment by GiudGiud** [comment]
@cpgr the AD page is linked to from some of the systems using AD
https://mooseframework.inl.gov/syntax/Kernels/ or https://mooseframework.inl.gov/syntax/Materials/ for example

**Comment by cpgr** [comment]
> @cpgr the AD page is linked to from some of the systems using AD
> https://mooseframework.inl.gov/syntax/Kernels/ or https://mooseframework.inl.gov/syntax/Materials/ for example

Thanks! I was wondering how anyone would find this.

**Comment by cpgr** [comment]
> @cpgr I was wondering if you looked at our porous Navier-Stokes slides at https://mooseframework.inl.gov/slides/index.html#/ whether you could tell me what the difference between that and the equations solved by the `porous_flow` module is? I'm hoping there is a difference haha, and if so we should think about documenting that somewhere so our users know where to go for what application/modeling-assumptions is relevant to them.

The governing equations for the `porous_flow` module are outlined here -https://mooseframework.inl.gov/modules/porous_flow/governing_equations.html

They are basically conservation of mass of a fluid component + Darcy's law.

I think the main difference would be that this module is set up for more classical reservoir engineering cases, and can model an arbitrary number of fluid phases (most common use case would be 2 I imagine - a gas phase and a liquid phase), but each phase can have an arbitrary number of fluid components (eg, the liquid can be water + salt + dissolved gas component etc). 

The FV stuff here (which is basically an FV equivalent to the FE objects already in this module) is to make it much easier to model cases where there are discontinuities in one of the variables, such as gas saturation , due to the reservoir properties.

**Comment by lindsayad** [comment]
It's fun learning about this. A really dumb question that I haven't answered for myself yet: if you were to model a single phase single component problem, what would your nonlinear variable(s) be? Is it just the pressure?

**Comment by cpgr** [comment]
> 

Yes

**Comment by cpgr** [comment]
I've been using this for a little while now and it all seems to be working.

The only part I'm not certain on is in `Coupleable.C`. I needed to template `coupledDofValues` to template all the porous flow materials, so added
```
template <>
const GenericVariableValue<false> &
Coupleable::coupledGenericDofValue<false>(const std::string & var_name, unsigned int comp) const
{
  return coupledDofValues(var_name, comp);
}

template <>
const GenericVariableValue<true> &
Coupleable::coupledGenericDofValue<true>(const std::string & var_name, unsigned int comp) const
{
  return adCoupledValue(var_name, comp);
}
```

but I'm not sure on the AD one.

**Comment by lindsayad** [comment]
Yea that's probably not quite right. The dof values should return all the degree of freedom values in the solution vector associated with a given element, whereas the coupled value api's return the finite element solution evaluated at the element quadrature points

**Comment by cpgr** [comment]
I couldn't find an `adCoupledDofValues` method - before I try and make one can I check that I'm not missing anything?

**Comment by lindsayad** [comment]
you're not missing anything 😄

File: framework/src/interfaces/Coupleable.C
Comment (lindsayad) [inline framework/src/interfaces/Coupleable.C:1754]: I think you’ve got these swapped

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/interfaces/Coupleable.C
Comment (lindsayad) [reply framework/src/interfaces/Coupleable.C:1754]: In the two lines below

Comment (cpgr) [reply framework/src/interfaces/Coupleable.C:1754]: Hehehe - I originally had `if(_coupleabe_neighbor)` but swapped it to `if(!_coupleabe_neighbor)` to match the existing methods and forgot to swap which one got returned.

**Comment by GiudGiud** [comment]
@cpgr are you waiting for a review on this?

**Comment by cpgr** [comment]
> @cpgr are you waiting for a review on this?

Yeah, but no rush. This was to check that what I have done so far is ok. I know that there is a lot of work to make it work with all of the exiting PorousFlow objects, I just don't have time until next month to keep going. 

If you or @loganharbour could just take a quick look and let me know if there is anything that I am not doing right that would be great.

**Comment by GiudGiud** [comment]
Sounds good. I ll add it to my list and I think @lindsayad will want to have a look as well since this is FV

**Comment by lindsayad** [comment]
Ah I had no idea you were waiting on a review. Sorry @cpgr !

**Comment by lindsayad** [comment]
I should have taken the fact that you marked it as ready for review as a cue haha

**Comment by cpgr** [comment]
No worries guys!

File: framework/src/interfaces/Coupleable.C
Comment (lindsayad) [inline framework/src/interfaces/Coupleable.C:535]: This doesn't look right, or at least I don't think this would be what the user would expect. I think what you would like to be returning from this is an array where the value part of the dual number is the dof value and the derivative part of the dof value has a single value of one at the corresponding dof index. Is that right? If so, maybe I could work to help you implement that

Comment (lindsayad) [reply framework/src/interfaces/Coupleable.C:535]: Oh wait... you already have this... so I think this just needs to be
```c++
  return adCoupledDofValues(var_name, comp);
```
Not sure why I'm unable to make a suggestion here :frowning:

File: modules/porous_flow/src/materials/PorousFlowSingleComponentFluid.C
Comment (lindsayad) [inline modules/porous_flow/src/materials/PorousFlowSingleComponentFluid.C:210]: A general comment I have is that these dy_dx quantities will themselves be functions of the nonlinear degrees of freedom. I have been slowly trying to add more emphasis to the the 5-parameter AD overloads. Just something to maybe think about

**Review by lindsayad — CHANGES_REQUESTED** [review]
Overall this looks really good, and I'm sorry I've taken so long to review. We need to get you and Andy added to the CCB. We've been making a lot of changes to the FV system recently and I hope we still support the work you've done here. We'll see when we re-trigger testing

File: modules/porous_flow/test/tests/pressure_pulse/tests
Comment (lindsayad) [inline modules/porous_flow/test/tests/pressure_pulse/tests:208]: Might be good to see a Jacobian test as well

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/interfaces/Coupleable.C
Comment (cpgr) [reply framework/src/interfaces/Coupleable.C:535]: Oops!

File: modules/porous_flow/src/materials/PorousFlowSingleComponentFluid.C
Comment (cpgr) [reply modules/porous_flow/src/materials/PorousFlowSingleComponentFluid.C:210]: I'll have to take a look at these - I remember getting a GitHub notification about them but didn't really take a good look

**Comment by cpgr** [comment]
Something has obviously changed - the flux kernel I was using no longer works! Any hints you could give me?

**Comment by lindsayad** [comment]
Hmm I think with @grmnptr's recent changes to remove ghost cells this might have plunged your neighbor material property evaluations when you're on a boundary into oblivion. What I would suggest, if you're wiling to wait, is to let #22130 go in, and then I can come push a commit to your PR updating your flux kernel. Would that be ok? I know you've already waited a long time, so if you're eager to have this in, then I can probably come up with a patch now

**Comment by cpgr** [comment]
Waiting is fine!

**Comment by lindsayad** [comment]
Alrighty, I'll re-save this and come back once #22130 is merged

**Comment by lindsayad** [comment]
Ok #22130 is in. Want the follow-on #23233 to get through and with those two I have a pretty good idea about where I want to take this. I don't think it'll be a massive re-write like I initially thought

**Comment by cpgr** [comment]
Let me know if there is anything I can do!

**Comment by lindsayad** [comment]
You've done plenty @cpgr. It's up to us to make sure your hard work doesn't go to waste

**Comment by lindsayad** [comment]
Ok @cpgr I finally got around to it! Sorry it took forever! I rebased onto a more recent MOOSE commit and then added one of my own. The important change is that we've removed ghost cells; so you have to be careful to avoid indexing into data that logically corresponds to a ghost cell, e.g. a "neighbor" when you are on a boundary face. In order to circumvent this change, I added a new `FVFluxBC` class that mimics the `FVFluxKernel` class you created but makes sure to avoid that out-of-bounds indexing. 

I tried to read about the porous flow module a bit, but I think it would be best if you looked at that new BC class and decided whether there might be a way to generalize it to more phases like you have for the flux kernel. Anyway hopefully this is helpful and you can develop on top of this and we can get this merged soon!

**Comment by cpgr** [comment]
Thanks so much @lindsayad. I'll take a look over the next few days

**Comment by cpgr** [comment]
I'm trying to understand why making a new flux bc fixes this? Does a FVFluxKernel not evaluate on the boundary faces anymore, meaning that a flux in/out must be added using a BC?

**Comment by lindsayad** [comment]
If a FVFluxBC object exists on a boundary, then the FVFluxKernel does not execute on that boundary. Here is the logic
```c++
bool
FVFluxKernel::skipForBoundary(const FaceInfo & fi) const
{
  // Boundaries to avoid come first, since they are always obeyed
  if (avoidBoundary(fi))
    return true;

  // We're not on a boundary, so in practice we're not 'skipping'
  if (!onBoundary(fi))
    return false;

  // Blanket forcing on boundary
  if (_force_boundary_execution)
    return false;

  // Selected boundaries to force
  for (const auto bnd_to_force : _boundaries_to_force)
    if (fi.boundaryIDs().count(bnd_to_force))
      return false;

  // If we have flux bcs then we do skip
  const auto & flux_pr = _var.getFluxBCs(fi);
  if (flux_pr.first)
    return true;

  // If we don't have flux bcs *and* we do have dirichlet bcs then we don't skip. If we don't have
  // either then we assume natural boundary condition and we should skip
  return !_var.getDirichletBC(fi).first;
}
```

**Comment by lindsayad** [comment]
One would hope. The reason I added it is for cell gradients. Cell gradients use Green-Gauss for their calculation so they need face values. If a Dirichlet BC exists then it will use that as the face value; if not, then by default we will use a two term extrapolation to determine the boundary face value based on the cell gradient (so this is actually an implicit calculation) and the cell center value. So practically speaking this extrapolated value probably won't match the Dirichlet value exactly, but hopefully it's close and is only an error of O(h^2).

You are certainly welcome to remove that code if you are happy with the results as they are

**Comment by cpgr** [comment]
> One would hope. The reason I added it is for cell gradients. Cell gradients use Green-Gauss for their calculation so they need face values. If a Dirichlet BC exists then it will use that as the face value; if not, then by default we will use a two term extrapolation to determine the boundary face value based on the cell gradient (so this is actually an implicit calculation) and the cell center value. So practically speaking this extrapolated value probably won't match the Dirichlet value exactly, but hopefully it's close and is only an error of O(h^2).
> 
> You are certainly welcome to remove that code if you are happy with the results as they are

**Comment by cpgr** [comment]
oops - accidentally closed this!

**Comment by cpgr** [comment]
The missing markdown documentation I can fix. The failing `Modules parallel` test throws an error when it runs the `--n-threads=2` tests

```
A 'FVBoundaryCondition' object already exists with the name 'left_diri'.
```

Does this just need an
```
if (_tid == 0)
 add Dirichlet BC here
```

**Comment by lindsayad** [comment]
Yes; that was a stupid oversight by me.

**Review by lindsayad — CHANGES_REQUESTED** [review]
I'd still like to see a Jacobian test, but otherwise this is good to go as far as I'm concerned.

**Comment by cpgr** [comment]
> I'd still like to see a Jacobian test, but otherwise this is good to go as far as I'm concerned.

I just added a Jacobian test of the FV kernels

File: modules/porous_flow/test/tests/jacobian/tests
Comment (lindsayad) [inline modules/porous_flow/test/tests/jacobian/tests:1219]:
```suggestion
  [fv_flux]
```

File: modules/porous_flow/test/tests/jacobian/tests
Comment (lindsayad) [inline modules/porous_flow/test/tests/jacobian/tests:1225]:
```suggestion
    requirement = 'PorousFlow shall compute all Jacobian entries of FV Kernels for flux and mass conservation.'
```
The SQA tsars have told me that requirements should be complete sentences

File: modules/porous_flow/test/tests/jacobian/tests
Comment (lindsayad) [inline modules/porous_flow/test/tests/jacobian/tests:1228]:
```suggestion
  []
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Review by lindsayad — APPROVED** [review]
Above and beyond on that last commit @cpgr :smile:

**Comment by cpgr** [comment]
> Above and beyond on that last commit @cpgr 😄

`sed` and some search and replace!

**Comment by GiudGiud** [comment]
@lindsayad do you need an external review on your commit or are we below the "significance" threshold?

**Comment by lindsayad** [comment]
I don't think it's very significant, so I don't think it needs it. Feel free to review if you wish obviously, but I think we're good to move forward without more review

File: modules/porous_flow/src/fvbcs/FVPorousFlow1PhaseAdvectiveFluxBC.C
Comment (GiudGiud) [inline modules/porous_flow/src/fvbcs/FVPorousFlow1PhaseAdvectiveFluxBC.C]: how often do you do 2D @cpgr ?

this default isnt great for 2D, I d almost want to error if this is used in 2D

Comment (cpgr) [reply modules/porous_flow/src/fvbcs/FVPorousFlow1PhaseAdvectiveFluxBC.C]: Yeah, I just used the same default we have in the other porous flow objects that exist. 

I guess we could either not have a default, or do something fancier like check the mesh dimension and then set an appropriate value?

Comment (GiudGiud) [reply modules/porous_flow/src/fvbcs/FVPorousFlow1PhaseAdvectiveFluxBC.C]: both are an option. It s not tied to this object then, can be figured out later

**Review by GiudGiud — APPROVED** [review]

**Comment by GiudGiud** [comment]
@cpgr please let me know when you want this merged

**Comment by lindsayad** [comment]
I bet he's ready to have it merged now 😄

**Comment by cpgr** [comment]
Haha - what's a few extra days!

**Comment by cpgr** [comment]
Thanks @lindsayad for your help getting this through!

**Comment by lindsayad** [comment]
You're welcome! Thanks for your patience in the process!

**Pull request merged**
